ftrr/src/main/java/framgia/co/edu/ftrr/service/impl/UserServiceImpl.java

```diff
@@ -86,4 +100,81 @@ public User loadCurrentLoginUser() {
             return null;
         }
     }
+
+    @Override
+    @Transactional
+    public CustomPrincipal loadOrUpdateUser(UserWsmResponse userWsmResponse) {
+        try {
+            checkUserWsmResponse(userWsmResponse);
+            User user = userRepository.getOneByEmail(userWsmResponse.getEmail()).orElse(new User());
+            getValueFromWsmToUser(user, userWsmResponse);
+            user = userRepository.save(user);
+            return buildCustomPrincipal(user);
+        } catch (Exception e) {
+            logger.error("Error in loadOrUpdateUser: " + e.getMessage());
+            return null;
+        }
+    }
+
+    private void checkUserWsmResponse(UserWsmResponse userWsmResponse) {
+        Optional.ofNullable(userWsmResponse.getGroups()).ifPresent(groups -> {
+            List<Group> groupsAfterUpdate = new ArrayList<>();
+            groups.stream().forEach(group -> {
+                Group g = new Group();
+                g.setId(group.getId());
+                g.setName(group.getName());
+                if (group.getGroups() != null) {
+                    g.setParentPath(JSONArray.toJSONString(group.getGroups().stream()
+                            .map(Group::getId).collect(Collectors.toList())));
+                }
+
+                Example<Group> groupExample = Example.of(g);
+                if (!groupRepository.exists(groupExample)) {
+                    groupsAfterUpdate.add(groupRepository.save(g));
+                } else {
+                    groupsAfterUpdate.add(groupRepository.getOne(group.getId()));
+                }
+            });
+            userWsmResponse.setGroups(groupsAfterUpdate);
+        });
+        if (isExistsPosition(userWsmResponse.getPosition())) {
+            userWsmResponse.setPosition(positionRepository.getOne(userWsmResponse.getPosition().getId()));
+        } else {
+            userWsmResponse.setPosition(positionRepository.save(userWsmResponse.getPosition()));
+        }
+    }
+
+    private boolean isExistsPosition(Position position) {
+        Position positionToSearch = new Position();
+        positionToSearch.setId(position.getId());
+        positionToSearch.setName(position.getName());
+        positionToSearch.setPositionType(position.getPositionType());
+
+        Example<Position> positionExample = Example.of(positionToSearch);
+        return positionRepository.exists(positionExample);
+    }
+
+    private void getValueFromWsmToUser(User user, UserWsmResponse userWsmResponse) {
+        user.setEmail(userWsmResponse.getEmail());
+        user.setCode(userWsmResponse.getEmployeeCode());
+        user.setName(userWsmResponse.getName());
+        user.setGender(userWsmResponse.getGender().equals("female") ? Gender.FEMALE.getCode() : Gender.MALE.getCode());
```

Comment:
`"female"` get từ enum gender luôn chứ ko setting string như ri hì

với nên ép kiểu sang in hoa hoặc thường hết đi, chứ ko so sánh sợ ko map

Reply:
female là cái mà wsm trả về đó chị

Reply:
tồ lô,
`userWsmResponse.getGender()` mới là get giá trị của wsm trả về

còn `"female"` là giá trị em lấy so sánh, lõ wsm đẹp trời trả về Female thì việc em so sánh là heo ăn rồi

nên chị mới nói là ép in hoa hết hoặc thường hết cho `userWsmResponse.getGender()` rồi mới đi so sánh với value của gender của mình

mà cái gender mình để Female thì ép giá trị  `userWsmResponse.getGender()` cũng phải chữ in hoa đầu tiên hỷ, ko là cũng heo ăn